zbus/src/connection/builder.rs

```diff
@@ -266,6 +274,11 @@ impl<'a> Builder<'a> {
     /// interfaces available immediately after the connection is established. Typically, this is
     /// exactly what you'd want. Also in contrast to [`zbus::ObjectServer::at`], this method will
     /// replace any previously added interface with the same name at the same path.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if path cannot be parsed as a valid object path (i.e.: if
+    /// `ObjectPath::try_from(path)` fails).
```

Comment:
you still kept the internal detail. Please remove the `(..)` part. It's not needed.

Reply:
The part between the parenthesis is where the actually useful information is contained.

I was writing some code that calls this method, and wasn't quite sure when this could return and error (and therefore, how to handle it). I'm passing an object of type `ObjectPath`. A description saying _"Returns an error if path cannot be parsed as a valid object path"_ doesn't say anything useful, because it's not clear at all _when_ this can fail or _why_.

By specifying that this fails when `ObjectPath::try_from(path)` would fail, then it is clear that this call is likely infallible (because `T::try_from(t: T)` is usually infallible), so I can just use `.expect()` assuming that this does not fail.

---

TBH, taking a `TryInto<T>` as input in a function like this is an anti-pattern. It forces the API to have to represent the error even when it is not applicable, even when it doesn't make sense. It would make much more sense to take an `ObjectPath` as input, and require that the caller do the validation/conversion [if necessary] themselves.

However, this isn't about fixing the API; I'm just trying to make the existing API clearer so it's more obvious when errors need to be handled and how.

Reply:
I feel similarly about this API (#321)

Reply:
> because it's not clear at all _when_ this can fail or _why_.

As a user, you don't need to know. If you do, look at the generic constraints. User just needs to know that they need to pass a valid object path. If they don't, they'll get an error.

> TBH, taking a `TryInto<T>` as input in a function like this is an anti-pattern. It forces the API to have to represent the error even when it is not applicable, even when it doesn't make sense. It would make much more sense to take an `ObjectPath` as input, and require that the caller do the validation/conversion [if necessary] themselves.

I disagree and hence whey we've the generic API.

As we discussed before, I'm not against providing API that takes ready objects but the main API should be as easy to use as possible (#321).

Reply:
> > because it's not clear at all _when_ this can fail or _why_.
> 
> As a user, you don't need to know. If you do, look at the generic constraints. User just needs to know that they need to pass a valid object path. If they don't, they'll get an error.

I actually have never seen the "how" being documented in "Error" sections of fallible methods in the Rust world and I think that's the right thing to do.

Reply:
> User just needs to know that they need to pass a valid object path.

Yeah, and IMHO it can be best described by the type system, by requiring the argument to be of a certain concrete type, not the documentation.

> I'm not against providing API that takes ready objects

I thought #321 was rejected... But good to know. Anyway, this is getting off-topic :)

Reply:
> > User just needs to know that they need to pass a valid object path.
> 
> Yeah, and IMHO it can be best described by the type system, 

It is! The generics say what's the expected type.

> by requiring the argument to be of a certain concrete type, not the documentation.

I'm not the one pushing for docs here but I'm not against them either. We're enforcing compliance whether user has to create the types or we do it for them.

> > I'm not against providing API that takes ready objects
>
> I thought #321 was rejected... But good to know. Anyway, this is getting off-topic :)

Hmm.. actually it was. I confused with #403, where the suggestion is to add additional API.

Reply:
> where the suggestion is to add additional API.

#321 also suggests to _add_ API (it's in the title :smile:), but #403 is _very_ similar but better written, so I'm happy to see that my idea is not dead.